llvm/include/llvm/CAS/FileSystemCache.h

```diff
@@ -160,7 +161,7 @@ class FileSystemCache : public ThreadSafeRefCountedBase<FileSystemCache> {
   /// returning early on a symlink.
   Expected<LookupPathState> lookupRealPathPrefixFrom(
       LookupPathState State, RequestDirectoryEntryType RequestDirectoryEntry,
-      PreloadTreePathType &PreloadTreePath,
+      PreloadTreePathType PreloadTreePath,
```

Comment:
@dexonsmith this part of the change is a little bit subtle: currently we only do preload realpath once, but after this change it will be once after each symlink traversal (but still only if necessary because we don't already know the realpath). This lets us compute `ExpectNonCanonical` correctly.

Reply:
This was pretty subtle before you made the change, too :).

I am seeing a potential problem. `lookupRealPathPrefixFrom()` can be called multiple times for a single lookup. I think the current change means that the stat+open+fcntl in `PreloadTreePath` will now be called repeatedly, instead of just once, even when there's no case mismatch. I think this testcase should expose it:
```
/tmp/path/to/a -> b
/tmp/path/to/b -> c
/tmp/path/to/c
```
IIRC, the call to `getStatus("/tmp/path/to/a")` will trigger something like:
- `preloadRealPathPrefixFrom({"/", "tmp/path/to/a"}, ...)`: PreloadTreePath calls stat+open+fcntl and make all the entries for `/tmp/path/to/c`; preloadRealPathPrefixFrom will navigate to `/tmp/path/to` and `lstat` symlink `a`
- `preloadRealPathPrefixFrom({"/tmp/path/to", "b"}, ...)`: PreloadTreePath calls stat+open+fcntl and then do nothing with it (new and undesired behaviour with this patch, IIUC); preloadRealPathPrefixFrom will `lstat` the symlink `b`
- `preloadRealPathPrefixFrom({"/tmp/path/to", "c"}, ...)`: PreloadTreePath navigates to `c` and returns it

Maybe it makes sense to introduce a struct/class to use here instead of `PreloadTreePath`:
- Points at the function to call / `PreloadTreePath`.
- Tracks whether it has been called yet (so it's not called again unnecessarily).
- Maybe, stores some extra information / state acquired during the first call to pass forward to future calls.

I'm wondering actually if changing the layering/callbacks/etc. entirely would make this much less subtle. Instead of passing around a bunch of possible-nullptr-callbacks, pass a reference to an object.

E.g.:
```
// In FilesystemCache.h
class FilesystemDiscoveryInstance {
  virtual void anchor() = 0;

public:
  virtual Expected<DirectoryEntry *> requestDirectoryEntry(...) = 0;
  virtual Error(DirectoryEntry &) requestSymlinkTarget(...) = 0;
  virtual void preloadRealPath(...) {}
  virtual void trackNonRealPathEntry(...) {}
};

class FileSystemCache {
  Expected<DirectoryEntry *>
  lookupPath(StringRef Path, DirectoryEntry &WorkingDirectory,
             FilesystemDiscoveryInstance &DI, bool FollowSymlinks);
};

// In CacheOnDiskFileSystem.cpp
namespace {
class DiscoveryInstance final : public FilesystemDiscoveryInstance {
public:
  Expected<DirectoryEntry *> requestDirectoryEntry(...) override;
  Error(DirectoryEntry &) requestSymlinkTarget(...) override;
  void preloadRealPath(...) override;
  void trackNonRealPathEntry(...) override;

  CachingOnDiskFileSystemImpl &FS;
  
  // Filled in on call to preloadRealPath(); contains the real path.
  DirectoryEntry *EntryForRealPath = nullptr;
  // From constructor; whether trackNonRealPathEntry is a no-op.
  bool IsTrackingStats = false;
  // Other useful state...
};

class CachingOnDiskFileSystemImpl ... {
  friend class DiscoveryInstance;
};

CachingOnDiskFileSystemImpl::lookupPath(
    StringRef Path, bool FollowSymlinks, bool LookupOnDisk,
    function_ref<void(FileSystemCache::DirectoryEntry &)>
        TrackNonRealPathEntries) {
  DiscoveryInstance DI(*this, LookupOnDisk, ...);
  Expected<DirectoryEntry *> ExpectedEntry =
     Cache.lookupPath(Path, DI, FollowSymlinks);
  // ...
}
```

Reply:
> I am seeing a potential problem. lookupRealPathPrefixFrom() can be called multiple times for a single lookup.

Right, yeah this isn't ideal.

> FilesystemDiscoveryInstance

This seems like a promising direction, I'll give it a shot, thanks!